### ISSUE TITLE
Wrong liblouis translations

### ISSUE THREAD

**Issue by egli** [sbsdev/daisyproducer#12]
The white list is generated by some Python code that compares the result of a liblouis translation to s set of known-good translations. The actual translation is then done by some Java code that uses this generated white list table.

For whatever reason the Java code invokes liblouis with different tables that the Python code that generates the white list. On the new production machine (with liblouis 3.2) this seems to make a difference

``` shell
eglic@xmlp01:/srv/xmlp-01.sbszh.ch/daisyproducer$ echo zwischenlandend | lou_trace sbs-wordsplit.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-g2-core.mod,sbs-special.mod
ZWL+DCD
1.	always	zwischen	1356-2456
2.	lowercase	l	123
3.	begmidword	an	235
4.	lowercase	d	145
5.	always	en	14
6.	lowercase	d	145
eglic@xmlp01:/srv/xmlp-01.sbszh.ch/daisyproducer$ echo zwischenlandend | lou_trace sbs.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-core.mod,sbs-de-g2-white.mod,sbs-de-g2-core.mod,sbs-de-hyph-none.mod,sbs-de-accents-ch.mod,sbs-special.mod
ZWL+DEND
1.	always	zwischen	1356-2456
2.	begmidword	lande	123-235-E-145-15-F
3.	lowercase	n	1345
4.	lowercase	d	145
5.	pass2	@E	?
6.	pass2	@F	?
```

while on the old machine (liblouis 2.5.2) it results in the same

``` shell
eglic@xmlp01:~$ echo zwischenlandend | lou_translate sbs-wordsplit.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-g2-core.mod,sbs-special.mod
ZWL+DCD
eglic@xmlp01:~$ echo zwischenlandend | lou_translate sbs.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-core.mod,sbs-de-g2-white.mod,sbs-de-g2-core.mod,sbs-de-hyph-none.mod,sbs-de-accents-ch.mod,sbs-special.mod
ZWL+DCD
```

Hm, maybe that was a bug in liblouis that is now fixed

``` shell
eglic@xmlp01:~$ lou_trace sbs-wordsplit.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-g2-core.mod,sbs-special.modzwischenlandend
ZWL+DCD
1.	always	zwischen	1356-2456
2.	lowercase	l	123
3.	begmidword	an	235
4.	lowercase	d	145
5.	always	en	14
6.	lowercase	d	145
^C
eglic@xmlp01:~$ lou_trace sbs.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-core.mod,sbs-de-g2-white.mod,sbs-de-g2-core.mod,sbs-de-hyph-none.mod,sbs-de-accents-ch.mod,sbs-special.modzwischenlandend
ZWL+DCD
1.	always	zwischen	1356-2456
2.	lowercase	l	123
3.	begmidword	an	235
4.	lowercase	d	145
5.	always	en	14
6.	lowercase	d	145
```
``` shell
eglic@wharton:~$ ssh xmlp
eglic@xmlp01:~$ echo zwischenlandend | lou_translate sbs.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-core.mod,sbs-de-g2-white.mod,sbs-de-g2-core.mod,sbs-de-hyph-none.mod,sbs-de-accents-ch.mod,sbs-special.mod
ZWL+DCD
eglic@xmlp01:~$ logout
eglic@wharton:~$ ssh xmlp-01
Last login: Fri Jun 16 16:05:07 2017 from 10.0.41.2
eglic@xmlp01:~$ echo zwischenlandend | lou_translate sbs.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-core.mod,sbs-de-g2-white.mod,sbs-de-g2-core.mod,sbs-de-hyph-none.mod,sbs-de-accents-ch.mod,sbs-special.mod
ZWL+DEND
```

**Comment by egli** [comment]
`git bisect` indicates that liblouis/liblouis@1f3c08b35f1d3118532b7b56b50d6f77c12a4e6e is the first bad commit

```
commit 1f3c08b35f1d3118532b7b56b50d6f77c12a4e6e
Author: Mike Gray <mgray@aph.org>
Date:   Fri Nov 11 11:09:07 2016 -0500
    Added empmatchbefore and empmatchafter prefix opcodes and updated UEB tables to use them.
    Allow capitals in checkEmphasisChange.
```

**Comment by egli** [comment]
Can't really tell what change in liblouis/liblouis@1f3c08b could cause this regression.

**Comment by egli** [comment]
The other thing is of course that the Python code invokes liblouis with different tables that the Java code. Some of it might be intentional, some might be a bug that didn't matter before

``` diff
diff -c /home/eglic/src/liblouis/python_tables.txt /home/eglic/src/liblouis/java_tables.txt
*** /home/eglic/src/liblouis/python_tables.txt	2017-06-19 16:11:04.657017083 +0200
--- /home/eglic/src/liblouis/java_tables.txt	2017-06-19 16:11:37.771612249 +0200
***************
*** 1,4 ****
! sbs-wordsplit.dis
  sbs-de-core6.cti
  sbs-de-accents.cti
  sbs-special.cti
--- 1,4 ----
! sbs.dis
  sbs-de-core6.cti
  sbs-de-accents.cti
  sbs-special.cti
***************
*** 6,10 ****
--- 6,14 ----
  sbs-de-letsign.mod
  sbs-numsign.mod
  sbs-litdigit-upper.mod
+ sbs-de-core.mod
+ sbs-de-g2-white.mod
  sbs-de-g2-core.mod
+ sbs-de-hyph-none.mod
+ sbs-de-accents-ch.mod
  sbs-special.mod

Diff finished.  Mon Jun 19 16:11:51 2017
```

**Comment by egli** [comment]
When invoking `lou_translate` w/o the `sbs-de-core.mod` table we get the right result:

``` shell
eglic@xmlp01:~$ echo zwischenlandend | lou_trace sbs.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-core.mod,sbs-de-g2-core.mod,sbs-de-hyph-none.mod,sbs-de-accents-ch.mod,sbs-special.modZWL+DEND
1.	always	zwischen	1356-2456
2.	begmidword	lande	123-235-E-145-15-F
3.	lowercase	n	1345
4.	lowercase	d	145
5.	pass2	@E	?
6.	pass2	@F	?
eglic@xmlp01:~$ echo zwischenlandend | lou_trace sbs.dis,sbs-de-core6.cti,sbs-de-accents.cti,sbs-special.cti,sbs-whitespace.mod,sbs-de-letsign.mod,sbs-numsign.mod,sbs-litdigit-upper.mod,sbs-de-g2-core.mod,sbs-de-hyph-none.mod,sbs-de-accents-ch.mod,sbs-special.mod
ZWL+DCD
1.	always	zwischen	1356-2456
2.	lowercase	l	123
3.	begmidword	an	235
4.	lowercase	d	145
5.	always	en	14
6.	lowercase	d	145

**Comment by egli** [comment]
See also liblouis/liblouis#369

**Comment by egli** [comment]
For now we just downgrade liblouis to an old version

**Comment by egli** [comment]
This problem will (hopefully) go away when we transition to a table that is generated by the maketable utility.

**Comment by bertfrees** [comment]
I think the preferred solution would be to fix [liblouis#369](https://github.com/liblouis/liblouis/issues/369)/[liblouis#105](https://github.com/liblouis/liblouis/issues/105).

The other solution, which you chose, is to fix the SBS table. Note that you should actually verify that after the big table refactoring it actually has less classes that the maximum allowed number. (As a matter of fact, we should do that check in liblouis: see [liblouis#105](https://github.com/liblouis/liblouis/issues/105#issuecomment-322435370).)
